File: app/models/cotizacion.py
from pydantic import BaseModel, EmailStr, field_validator
from typing import List, Optional


class ServicioItem(BaseModel):
    code:     str
    name:     str
    norma:    Optional[str]   = None
    sub:      Optional[str]   = None
    muestras: Optional[int]   = None
    precio:   Optional[float] = None


class UbicacionModel(BaseModel):
    lat:     str
    lng:     str
    address: Optional[str] = None


class CotizacionRequest(BaseModel):
    nombre:            str
    correo:            EmailStr
    empresa:           str
    telefono:          str
    rtn:               Optional[str] = None
    nombreProyecto:    str
    direccionProyecto: Optional[str] = None
    descripcion:       Optional[str] = None
    servicios:         List[ServicioItem]
    ubicacion:         Optional[UbicacionModel] = None

    @field_validator("nombre")
    @classmethod
    def nombre_no_vacio(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("El nombre no puede estar vacío")
        return v.strip()

    @field_validator("empresa")
    @classmethod
    def empresa_no_vacia(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("La empresa o institución no puede estar vacía")
        return v.strip()

    @field_validator("telefono")
    @classmethod
    def telefono_no_vacio(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("El teléfono no puede estar vacío")
        return v.strip()

    @field_validator("nombreProyecto")
    @classmethod
    def nombre_proyecto_no_vacio(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("El nombre del proyecto no puede estar vacío")
        return v.strip()

    @field_validator("servicios")
    @classmethod
    def al_menos_un_servicio(cls, v: List[ServicioItem]) -> List[ServicioItem]:
        if not v:
            raise ValueError("Debe seleccionar al menos un servicio")
        return v


class CotizacionResponse(BaseModel):
    ok:      bool
    message: str
    fila:    Optional[int] = None

File: src/validators/cotizacion.js
'use strict';
const { z } = require('zod');

const ServicioItem = z.object({
  code:     z.string().min(1),
  name:     z.string().min(1),
  norma:    z.string().optional().nullable(),
  sub:      z.string().optional().nullable(),
  muestras: z.number().int().positive().optional().nullable(),
  precio:   z.number().nonnegative().optional().nullable(),
});

const UbicacionModel = z.object({
  lat:     z.string(),
  lng:     z.string(),
  address: z.string().optional().nullable(),
});

const CotizacionRequest = z.object({
  nombre:            z.string().trim().min(1, 'El nombre no puede estar vacío'),
  correo:            z.string().email('Correo inválido'),
  empresa:           z.string().trim().min(1, 'La empresa no puede estar vacía'),
  telefono:          z.string().trim().min(1, 'El teléfono no puede estar vacío'),
  rtn:               z.string().optional().nullable(),
  nombreProyecto:    z.string().trim().min(1, 'El nombre del proyecto no puede estar vacío'),
  direccionProyecto: z.string().optional().nullable(),
  descripcion:       z.string().optional().nullable(),
  servicios:         z.array(ServicioItem).min(1, 'Debe seleccionar al menos un servicio'),
  ubicacion:         UbicacionModel.optional().nullable(),
});

module.exports = { CotizacionRequest };
